docs/coding-guidelines/mono-code-guide.md

```diff
@@ -0,0 +1,253 @@
+
+* `sgen` depends on `eglib` and `utils`
+
+* `metadata` depends on all of the above (but note that some Boehm-GC exists that should not depend on `sgen`)
+
+* `mini` depends on all of the above
+
+* `mini/interp` depends on all of the above
+
+* `components` can use functions from all of the above provided they're marked with
+  `MONO_COMPONENT_API`, see [../design/mono/components.md](../design/mono/components.md)
+
+The main distinction between `metadata` and `utils` is that utils code should not assume that it is
+part of an executing .NET runtime - anything that loads types, creates objects, etc does not belong
+in utils but in metadata.
+
+The `mini` directory contains execution engines.  If the execution engine has to provide some
+functionality to `metadata` it generally does so by installing some callback that is invoked by
+`metadata`.  For example, `mini` knows how to unwind exceptions and perform stack walks - while
+`metadata` decides *when*_* to unwind or perform a stackwalk.  To coordinate, `metadata` exposes an
+API to install hooks and `mini` provides the implementations.
+
+## Error handling
+
+New code should prefer to use the `MonoError` functions.
+
+* A non-public-API function should take a `MonoError *` argument.
+* In case of an error the function should call one of the `mono_error_set_` functions from [../../src/mono/mono/utils/mono-error-internals.h](../../src/mono/mono/utils/mono-error-internals.h)
+* Inside the runtime check if there was an error by calling `is_ok (error)`, `mono_error_assert_ok (error)`, `goto_if_nok` or `return_if_nok`/`return_val_if_nok`
+* If there is an error and you're handling it, call `mono_error_cleanup (error)` to dispose of the resources.
+* `MonoError*` is generally one-shot: after it's been cleaned up it need to be re-inited with `mono_error_init_reuse`, but this is discouraged.
+* Instead if you intend to deal with an error, use `ERROR_DECL (local_error)` then call the
+  possibly-failing function and pass it `local_error`, then call `mono_error_assert_ok
+  (local_error)` if it "can't fail" or `mono_error_cleanup (local_error)` if you want to ignore the
+  failure.
+
+## Managed Exceptions
+
+New code should generally not deal with `MonoException*`, use `MonoError*` instead.
+
+New code should avoid using `mono_error_set_pending_exception` - it affects a thread-local flag in
+a way that is not obvious to the caller of your function and may trample existing in-flight
+exceptions and make your code fail in unexpected ways.
+
+There are two circumstances when you might need to call `mono_error_set_pending_exception`:
+
+1. You're working with a public Mono API that sets a pending exception
+2. You're implementing an icall, but can't use `HANDLES()` (see the [internal calls](#internal-calls) section below)
+
+## Internal calls <a name="internal-calls"></a>
+
+Prefer P/Invokes or QCalls over internal calls.  That is, if your function only takes arguments that
```

Comment:
do we have a doc about QCalls that we could link here?

Reply:
Not specifically for Mono.  I actually don't know what the happy path for adding a QCall on Mono looks like right now.  **Help wanted**